Provide new "cid" sub-command.

kevina opened the pull request:
Implements #5357.

~Still needs tests.~

Closes #5229. Closes #5357.

Stebalien commented on core/commands/cid.go:
```diff
@@ -0,0 +1,307 @@
+` + cidutil.FormatRef,
+	},
+	Arguments: []cmdkit.Argument{
+		cmdkit.StringArg("cid", true, true, "Cids to format."),
```
Can we use `WithDefault("%s")`?

kevina replied:
You meant for the `-f` option?  If it is omitted that is the default (see below), I just didn't use WithDefault.  The idea was the `-f` is optional and if its not specified it will just return the CIDs.

Stebalien replied:
It's nice from a documentation standpoint.

kevina replied:
No Problem.

Stebalien commented on core/commands/cid.go:
```diff
@@ -0,0 +1,307 @@
+		case "":
+			opts.fmtStr = "%s"
+		case "prefix":
+			opts.fmtStr = "%P"
```
This feels a bit magical. What's the motivation? We could also just add a `ipfs cid prefix` command (alias of `ipfs cid format -f '%P' ...`).

kevina replied:
It was there in the original `cid-fmt` utility and copied over here.  For the `cid-fmt` utility is was the original use case (to get information about cids).  I am not tied to keeping it in this utility.

Stebalien replied:
I'd remove it for now. We can consider either a `prefix` or a `show` command later but we probably don't *need* them now.

kevina replied:
Will do.

Stebalien commented on core/commands/cid.go:
```diff
@@ -0,0 +1,307 @@
+			opts.newBase = mbase.Encoding(-1)
+		}
+
+		res, err := formatCids(req.Arguments, opts)
```
I'd stream these results back. Users may want to feed quite a few CIDs into this command.

kevina replied:
I thought about that but didn't see the benefit.  This is a client side utility that will never block and is should be very fast, so the only benefit will be memory usage.  In addition the number of cid's passed in is limited by the length of command line so memory usage will never get out of control.

Stebalien replied:
Ah. Well, actually, arguments can be passed in on stdin (newline separated)... That's why I brought this up. Basically, a user may run something like `ipfs refs -u -r QmThing | ipfs cid format -f '%c' | sort -u`.

kevina replied:
I think you have to do something special to make that work.  Because when I tried it I get `Error: argument "cid" is required`.  I actually think this it is a bit of a misfeature to take the arguments from stdin without having to pass in a special flag, but that something to debate later.

warpfork replied:
Fwiw, I agree it's a misfeature to take arguments from stdin without a flag -- I think it's pretty widespread convention now to use a single dash "`-`" argument to mean "read stdin instead of file".  (And that would make just as much sense for "read stdin instead of literal".)

Stebalien commented:
@alanshaw this adds a new `ipfs cid` command. I'm fine with keeping this command "go specific" (it's really just a utility command) but thought you might have some input.

Basically, it makes it easy to inspect CIDs and convert between various formats (bases, versions, etc.).

Stebalien commented on core/commands/cid.go:
```diff
@@ -0,0 +1,314 @@
+			emit(str, err)
+		}
+	}()
+	resp.Emit(out)
```
You can just call Emit on each CID individually (no need for the goroutine or the channel).

Stebalien commented:
EnableStdin isn't broken, it's just that:

1. The documentation is all missing.
2. The API sucks.

Basically, you'll have to:

1. Iterate over all the arguments in `Arguments`.
2. Iterate over the arguments from `BodyArgs()`.

You can also just call `ParseBodyArgs()` to move all the arguments from the body into `Arguments`.

kevina commented:
> You can also just call ParseBodyArgs() to move all the arguments from the body into Arguments.

That will prevent streaming.

Stebalien commented:
> That will prevent streaming.

It will. The important part here is *allowing* ourselves to stream at some point in the future without breaking the API.

Stebalien commented on core/commands/cid.go:
```diff
@@ -0,0 +1,336 @@
+		cmds.Text: cmds.MakeEncoder(func(req *cmds.Request, w io.Writer, val0 interface{}) error {
+			prefixes, _ := req.Options["prefix"].(bool)
+			numeric, _ := req.Options["numeric"].(bool)
+			val := val0.([]CodeAndName)
```
For now, you'll need to check if the type is correct. See: https://github.com/ipfs/go-ipfs-cmds/issues/65. Basically, we end up feeding errors to this function as well.

(search the code for examples on how to do this).

Stebalien commented on core/commands/commands.go:
```diff
@@ -152,3 +153,46 @@ func unwrapOutput(i interface{}) (interface{}, error) {
+
+// streamRes is a helper function to stream results, that possibly
+// contain with non-fatal, the helper function is allowed to panic on
+// internal errors
```
Needs some proof-reading :).

Stebalien commented on core/commands/commands.go:
```diff
@@ -152,3 +153,46 @@ func unwrapOutput(i interface{}) (interface{}, error) {
+// streamRes is a helper function to stream results, that possibly
+// contain with non-fatal, the helper function is allowed to panic on
+// internal errors
+func streamRes(procVal func(interface{}, io.Writer) nonFatalError) func(*cmds.Request, cmds.ResponseEmitter) cmds.ResponseEmitter {
```
Can we call this `streamResults`? We tend to a bit *too* curt.

Stebalien commented on core/commands/commands.go:
```diff
@@ -152,3 +153,46 @@ func unwrapOutput(i interface{}) (interface{}, error) {
 	return <-ch, nil
 }
+
+type nonFatalError string
```
Why can't we just use an `error`?

Stebalien commented on core/commands/cid.go:
```diff
@@ -0,0 +1,336 @@
+		if !ok {
+			break
+		}
+		emit := func(fmtd string, err error) {
```
Let's lift this out of the loop (I don't trust go).

kevina replied:
It's in the loop as it is a closure.  I do it that way to avoid having to pass is the cidStr variable.

kevina replied:
I just eliminated the need for the emit function.

Stebalien commented on core/commands/cid.go:
```diff
@@ -0,0 +1,336 @@
+			if err != nil {
+				res.ErrorMsg = err.Error()
+			}
+			resp.Emit(res)
```
We should handle this error (the client may have gone away).

kevina replied:
I'm not sure what I should do.  Should I log the error and then just return?

kevina replied:
I made at attempt at handling this, please check that it is correct.

kevina commented:
> Is there any way to simplify this? We're introducing multiple new abstractions.

Most (all) of the abstractions are lifted from the `filestore ls` code when I am doing nearly the same thing.  Most of the completely comes from the fact that I believe fairly strongly that we should not abort on non-fatal errors, especially when streaming.

kevina commented:
> Is there any way to simplify this? We're introducing multiple new abstractions.

On second thought.  I am not 100% sure what you mean.  I thought you where talking about the handling of the `cid fotmat` command and streaming?  Is there anything else you are referring to?

alanshaw commented:
Hey everyone! I've tried to replicate the functionalty here as closely as possible for the `jsipfs cid` command. I'd appreciate it if you're able to spare a few minutes to look over https://github.com/ipfs-shipyard/js-cid-tool (if not the code, the README will do) to check I've captured the intentions correctly.

thank you ❤️

kevina commented:
@alanshaw I looked over the README and I think the tool captures the intentions about the same.  It was never really my intention for these commands to become standardized except for perhaps the `base32` command.

Note that some minor details of the behavior of these commands may change over time.

kevina commented:
@Stebalien if this is too complicated, I can separate out the part that implements proper streaming into a separate p.r. as I think that is where most of the complexity your seeing is coming from...

alanshaw commented:
Thanks @kevina 

> It was never really my intention for these commands to become standardized

Understood, I haven't added this to any `interface-*` repos. I'm happy for this to just be a command line tool. I think it's important that `jsipfs` users also have it which is why I created it. Plus it was pretty quick and easy to put together.

Stebalien commented:
@kevina I was just waiting on *some* second signoff. Now that we have one, this should be good to go. Unfortunately, it'll need a rebase. I'd squash first as doing this commit-by-commit is painful (I tried, but there were significant changes to the commands lib).

kevina commented:
@Stebalien I rebase later today.

kevina commented:
@Stebalien actually I already rebased and fixed my code to use the new commands library as indicated in an earlier comment.  I just rebased again and fixed a dep/import problem.

This should be good to go.  There is one failed test in Jenkins, but I don't think that is related.

kevina commented:
And it looks like rebasing caused GitHub to dismiss the earlier approving reviews...

kevina commented:
@Stebalien Jenkins is now green.

Stebalien commented:
> And it looks like rebasing caused GitHub to dismiss the earlier approving reviews...

Yeah, I just turned that off. Sorry.
